src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -238,28 +243,190 @@ private SNITCPHandle CreateTcpHandle(string fullServerName, long timerExpire, ob
                     return null;
                 }
             }
-            else if (serverAndPortParts.Length > 2)
+            else
             {
                 SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, SNICommon.InvalidConnStringError, string.Empty);
                 return null;
             }
+        }
+
+        /// <summary>
+        /// Finds IPAddress by given hostname.
+        /// </summary>
+        /// <param name="hostname">hostname string</param>
+        /// <param name="hostIPAddress">host IPAddress found by hostname string</param>
+        /// <param name="family">address family. Default value is AddressFamily.InterNetwork</param>
+        /// <returns>bool value showing if operation is successful</returns>
+        private static bool GetIPAdress(string hostname, out IPAddress hostIPAddress, AddressFamily family = AddressFamily.InterNetwork)
+        {
+            hostIPAddress = null;
+
+            if (String.IsNullOrWhiteSpace(hostname))
+            {
+                return false;
+            }
+
+            try
+            {
+                Task<IPAddress[]> addressTask = Dns.GetHostAddressesAsync(hostname);
+                IPAddress[] addresses = null;
+                if (!addressTask.Wait(1000) || (addresses = addressTask.Result) == null || addresses.Length == 0)
+                {
+                    return false;
+                }
+
+                List<IPAddress> addressList = new List<IPAddress>(addresses);
+                hostIPAddress = addressList.First(a => a.AddressFamily == family);
+                if (hostIPAddress == null)
+                {
+                    return false;
+                }
+            }
+            catch
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Sends CLNT_UCAST_INST request for given instance name to SQL Sever Browser, and receive SVR_RESP from the Browser.
+        /// </summary>
+        /// <param name="browserEP">SQL Sever Browser IPEndPoint</param>
+        /// <param name="instanceName">instance name for CLNT_UCAST_INST request</param>
+        /// <param name="responsePacket">SVR_RESP packets from SQL Sever Browser</param>
+        /// <returns>bool value showing if operation is successful</returns>
+        private static bool Send_CLNT_UCAST_INST(IPEndPoint browserEP, string instanceName, out byte[] responsePacket)
+        {
+            responsePacket = null;
+
+            if (browserEP == null || String.IsNullOrWhiteSpace(instanceName))
+            {
+                return false;
+            }
+
+            List<byte> byteList = new List<byte>(new byte[] { 0x04 });
+            byteList.AddRange(Encoding.ASCII.GetBytes(instanceName));
+            byte[] requestPacket = byteList.ToArray();
```

Comment:
There are a bunch of unnecessary allocations here. The above could be simplified as something like:

```c#
int byteCount = Encoding.ASCII.GetByteCount(instanceName);
byte[] requestPacket = new byte[byteCount + 1];
requestPacket[0] = 0x04;
Encoding.ASCII.GetBytes(instanceName, 0, instanceName.Length, requestPacket, 1);
```